src/Vault.sol

```diff
@@ -780,4 +782,48 @@ contract Vault {
     function getLengths() external view returns (uint256, uint256, uint256, uint256) {
         return (erc20Stored.length, erc721Stored.length, erc721TokenIds.length, erc1155Stored.length);
     }
+
+    /*///////////////////////////////////////////////////////////////
+                    ASSET MANAGEMENT LOGIC
+    ///////////////////////////////////////////////////////////////*/
+
+    function vaultManagementAction(address _actionHandler, bytes calldata _actionData) public onlyOwner {
+        require(IMainRegistry(registryAddress).isActionAllowlisted(_actionHandler), "VL_VMA: Action is not allowlisted");
+
+        (actionAssetsData memory outgoing_, actionAssetsData memory incoming_,) =
+            abi.decode(_actionData, (actionAssetsData, actionAssetsData, bytes));
+
+        // account preActionBalances
+        uint256 incomingLength = incoming_.assets.length;
+        for (uint256 i; i < incomingLength;) {
+            incoming_.preActionBalances[i] = IERC20(incoming_.assets[i]).balanceOf(address(this));
+            unchecked {
+                ++i;
+            }
+        }
+
+        // withdraw to actionHandlerå
+        for (uint256 i; i < outgoing_.assets.length;) {
+            outgoing_.preActionBalances[i] = IERC20(outgoing_.assets[i]).balanceOf(address(this));
+            _withdrawERC20(_actionHandler, outgoing_.assets[i], outgoing_.assetAmounts[i]);
+            unchecked {
+                ++i;
+            }
+        }
+
+        // execute Action
+        IActionBase(_actionHandler).executeAction(address(this), _actionData);
+
+        for (uint256 i; i < incomingLength;) {
+            _depositERC20(_actionHandler, incoming_.assets[i], incoming_.assetAmounts[i]);
```

Comment:
How is `incoming_.assetAmounts[i]` defined here? This should be the return values of executeAction to know how much to deposit

Reply:
+ also check why tests didn't pick this up.

Reply:
Tests didn't pick this up because mocks make sure exact amounts are swapped into the actionHandler